[BEAM-2603] add meter interface and implementation.

unsleepy22 opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

unsleepy22 commented:
R: @aviemzur @bjchambers @peihe could you help review this PR?

peihe commented:
Reviewed 17 of 20 files at r1.
Review status: 17 of 20 files reviewed at latest revision, 10 unresolved discussions.

---

*[pom.xml, line 1141 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-KoutXohx7wlWJtkKWRO:-KoutXohx7wlWJtkKWRP:b-fzt61k) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/pom.xml#L1141)):*
> ```XML
>         <groupId>io.dropwizard.metrics</groupId>
>         <artifactId>metrics-core</artifactId>
>         <version>3.1.2</version>
> ```

add dropwizard.version

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java, line 56 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov0pL6XjmS7L1JXzo6:-Kov0pL6XjmS7L1JXzo7:b2a95y0) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java#L56)):*
> ```Java
>   }
> 
>   public void mark() {
> ```

@Override?

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 29 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov2GfVpz7vG1c8dIjq:-Kov2GfVpz7vG1c8dIjr:bevqtfb) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L29)):*
>  MetricData

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 50 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov10vUs1Vjq5-vWMge:-Kov10vUs1Vjq5-vWMgf:b1k5zkq) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L50)):*
> ```Java
>   }
> 
>   public MeterData combine(MeterData other) {
> ```

why need combine() and merge()?
Could we only keep one

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 54 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov-opQhPCdbEWE2qXA:-Kov-opQhPCdbEWE2qXB:bolgzjd) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L54)):*
> ```Java
>   }
> 
>   public static MeterData zero() {
> ```

change to:
public static final MeterData ZERO = create(0, 0, 0, 0, 0);

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java, line 212 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-KouucsNVGGqgqaLf_vp:-KouucsNVGGqgqaLf_vq:b8vj90w) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java#L212)):*
> ```Java
>       MetricsMap<MetricName, MeterCell> updates) {
>     for (Map.Entry<MetricName, MeterCell> counter : updates.entries()) {
>       //current.get(counter.getKey()).update(counter.getValue().getCumulative());
> ```

why commented out?

---

*[runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectMetricsTest.java, line 98 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-KouzIpaJ8_m6U7fb9jz:-KouzIpbe6gFHkfDvbO6:b13gawj) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectMetricsTest.java#L98)):*
> ```Java
>         ImmutableList.of(
>             MetricUpdate.create(MetricKey.create("step1", NAME3),
>                 MeterData.create(1, 1, 1, 1, 1)))
> ```

add assertions to verify

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java, line 169 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov4SsiiTilHkil53_S:-Kov4SsiiTilHkil53_T:b-wt863m) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java#L169)):*
> ```Java
>         flinkMeterCache.put(flinkMetricName, meter);
>       } else {
>         meter.markEvent(update.count());
> ```

Will that mark duplicate event, given update.count() is the accumulated number.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Meter.java, line 24 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kouxd6oj-SQbV5GoXcu:-Kouxd6oj-SQbV5GoXcv:bc85404) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Meter.java#L24)):*
> ```Java
> 
> /**
>  * Meter interface which corresponds to Meter in dropwizard metrics.
> ```

We don't need to mention dropwizard in Beam interface.

-> "A metric that reports ..." similar as Counter/Gauge 

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricData.java, line 24 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Koux9ODun3peysPOA0C:-Koux9ODun3peysPOA0D:bo7ajr2) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricData.java#L24)):*
> ```Java
> 
> /**
>  * base class of metric data.
> ```

Say more about MetricData

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

unsleepy22 commented:
Review status: 17 of 20 files reviewed at latest revision, 10 unresolved discussions.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java, line 56 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov0pL6XjmS7L1JXzo6:-KovnMTEsTmw13JE1tHv:b-896fix) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java#L56)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

@Override?
</blockquote></details>

Done.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 29 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov2GfVpz7vG1c8dIjq:-KovngKeEMC0LA7gk-ps:b-896fix) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L29)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

>  MetricData
</blockquote></details>

Done.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 50 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov10vUs1Vjq5-vWMge:-KovnKL98bHY5RVksWiz:b-896fix) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L50)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

why need combine() and merge()?
Could we only keep one
</blockquote></details>

Done.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 54 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov-opQhPCdbEWE2qXA:-KovnJidLbo8_UGCMeGl:b-896fix) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L54)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

change to:
public static final MeterData ZERO = create(0, 0, 0, 0, 0);
</blockquote></details>

Done.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java, line 212 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-KouucsNVGGqgqaLf_vp:-KovnIrEu70Ic3gKc4ZR:b-896fix) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java#L212)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

why commented out?
</blockquote></details>

Done.

---

*[runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectMetricsTest.java, line 98 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-KouzIpaJ8_m6U7fb9jz:-KovnHuJbPt2iMVMqQPJ:b-896fix) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectMetricsTest.java#L98)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

add assertions to verify
</blockquote></details>

Done.

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java, line 169 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov4SsiiTilHkil53_S:-Kovn80S-7pZ5FhBJKmp:b-h1gmch) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java#L169)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

Will that mark duplicate event, given update.count() is the accumulated number.
</blockquote></details>

Yep good point I missed it. I changed the implementation of FlinkMeter.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Meter.java, line 24 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kouxd6oj-SQbV5GoXcu:-Kovn75DJ6efgMAAYhqF:b-896fix) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Meter.java#L24)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

We don't need to mention dropwizard in Beam interface.

-> "A metric that reports ..." similar as Counter/Gauge 



</blockquote></details>

Done.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricData.java, line 24 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Koux9ODun3peysPOA0C:-Kovn6Ch6l8Zj44HQNrT:b-896fix) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricData.java#L24)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

Say more about MetricData
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

peihe commented:
Reviewed 1 of 20 files at r1, 11 of 11 files at r2.
Review status: 19 of 21 files reviewed at latest revision, 4 unresolved discussions.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java, line 212 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-KouucsNVGGqgqaLf_vp:-KoyyoCq2asggaWHJCVf:b-1zv5gv) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java#L212)):*
<details><summary><i>Previously, unsleepy22 (Cody) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

LGTM

and, this also means this line is not cover by unit tests. Could you add one?

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java, line 169 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov4SsiiTilHkil53_S:-KoyzDqfKwIEEhsQVcL4:bawujs0) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java#L169)):*
<details><summary><i>Previously, unsleepy22 (Cody) wrote…</i></summary><blockquote>

Yep good point I missed it. I changed the implementation of FlinkMeter.
</blockquote></details>

LGTM

is this covered by unit tests?

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java, line 255 at r2](https://reviewable.io:443/reviews/apache/beam/3555#-KoyxVXRiOYkh2EJB3rh:-KoyxVXSSLRQa7UHQIJo:b-kykflu) ([raw file](https://github.com/apache/beam/blob/8425b8708f2d25cc80878de29d2df74a4865fee5/runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java#L255)):*
> ```Java
>      */
>     @Override
>     public void markEvent(long l) {
> ```

change to "long n"

l looks too close to 1.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

unsleepy22 commented:
Review status: 19 of 21 files reviewed at latest revision, 4 unresolved discussions.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java, line 212 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-KouucsNVGGqgqaLf_vp:-Koz5ZhzWe42lQnfFwsy:b-dr2wj3) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java#L212)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

LGTM

and, this also means this line is not cover by unit tests. Could you add one?
</blockquote></details>

OK, will add. Note that since this PR is about Meter, I'll add meter test only.

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java, line 169 at r1](https://reviewable.io:443/reviews/apache/beam/3555#-Kov4SsiiTilHkil53_S:-Koz3ggPgu5c2tq5RxUv:b-896fix) ([raw file](https://github.com/apache/beam/blob/4df6630d20a3005e058e541806aae561d014ab7d/runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java#L169)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

LGTM

is this covered by unit tests?
</blockquote></details>

Done.

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java, line 255 at r2](https://reviewable.io:443/reviews/apache/beam/3555#-KoyxVXRiOYkh2EJB3rh:-Koz3hinNjjCsxTvohCK:b-896fix) ([raw file](https://github.com/apache/beam/blob/8425b8708f2d25cc80878de29d2df74a4865fee5/runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java#L255)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

change to "long n"

l looks too close to 1.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

peihe commented:
This looks good to me in general.

@aviemzur, could you take another look into the metrics aspect?

bjchambers commented:
Review status: 18 of 22 files reviewed at latest revision, 8 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1BnmeU5aYp435knSC:-Kp1BnmeU5aYp435knSD:b-9uyfeq) (no related file):*
This departs from the current approach to metrics. Rather than allowing a runner to choose how to report metrics, it pushes drop wizard into the cross-runner implementation. This has other affects too, such as making it impossible to support meters across only the successfully committed work.

A lot of the complexity seems to come from the trying to use drop wizard directly while also supporting the general case of metrics with committed values if a runner supports them.

I'd like to see a discussion on the dev@ list about whether moving towards the workers reporting metrics directly to a plugin. If we like this approach, it means we can remove all the cruft around combining/storing metrics in separate cells/etc. Assuming that makes sense, we should still think about whether we want that top level connection to be made directly to drop wizard or through a layer of indirection allowing other libraries.

---

*[pom.xml, line 1140 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp19tbbylCvL-_NJvd_:-Kp19tbbylCvL-_NJvda:b-whggvx) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/pom.xml#L1140)):*
> ```XML
> 
>       <dependency>
>         <groupId>io.dropwizard.metrics</groupId>
> ```

Why are we adding dropwizard as a top-level dependency?

---

*[runners/core-java/pom.xml, line 99 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1AA0Vvad8uhYwzMca:-Kp1AA0Vvad8uhYwzMcb:bl4kp1j) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/pom.xml#L99)):*
> ```XML
> 
>     <dependency>
>       <groupId>io.dropwizard.metrics</groupId>
> ```

Ditto -- dropwizard should be an implementation detail on those runners that report to dropwizard, correct?

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java, line 38 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1AGjycjkaup_Fr4lL:-Kp1AGjycjkaup_Fr4lM:bbv5jnb) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java#L38)):*
> ```Java
> 
>   private final DirtyState dirty = new DirtyState();
>   private transient com.codahale.metrics.Meter value = new com.codahale.metrics.Meter();
> ```

Wrapping Meter like this will make it impossible for any runner to implement meters without using a drop wizard Meter. I believe that also makes it impossible to support committed values. This part of the implementation should instead capture enough information about the reported rate to export information to drop wizard within specific runners.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java, line 73 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1AXj1Ga34Inr-wHGE:-Kp1AXj1Ga34Inr-wHGF:b-y750qf) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java#L73)):*
> ```Java
>         value.getFiveMinuteRate(),
>         value.getFifteenMinuteRate(),
>         value.getMeanRate());
> ```

How would this data be mergeable? If the times aren't properly aligned, we can't just add the rates together, correct?

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 49 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1Adm0vIK3FzCRiXe0:-Kp1Adm0vIK3FzCRiXe1:bskx4h) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L49)):*
> ```Java
>         m5() + other.m5(),
>         m15() + other.m15(),
>         mean() + other.mean());
> ```

Adding mean rates together isn't sufficient for combining.

Consider:

A MeterData from 7:00 to 10:00 (3 hours) with an average rate of 100 and
A MeterData from 9:00 to 10:00 (1 hour) with an average rate of 100.

Why is the output rate an average of 200? If we assume the rate was evenly distributed, then we have:

2 hours with a rate of 33/hour, and 1 hour with a rate of 133/hour, etc.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

kennknowles commented on runners/core-java/pom.xml:
```diff
@@ -95,6 +95,11 @@
     </dependency>
 
+    <dependency>
+      <groupId>io.dropwizard.metrics</groupId>
```
I think this should live in its own module that is opt-in separate from runners-core.

bjchambers commented:
Review status: 18 of 22 files reviewed at latest revision, 9 unresolved discussions.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 49 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1Adm0vIK3FzCRiXe0:-Kp1D_rWjqnD7CpKOrE0:b-ahop2r) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L49)):*
<details><summary><i>Previously, bjchambers (Ben Chambers) wrote…</i></summary><blockquote>

Adding mean rates together isn't sufficient for combining.

Consider:

A MeterData from 7:00 to 10:00 (3 hours) with an average rate of 100 and
A MeterData from 9:00 to 10:00 (1 hour) with an average rate of 100.

Why is the output rate an average of 200? If we assume the rate was evenly distributed, then we have:

2 hours with a rate of 33/hour, and 1 hour with a rate of 133/hour, etc.
</blockquote></details>

Another similar problem -- we may get two meter data from the same thread that include overlapping rates. Again, adding the average rates together may cause numerical problems such as double counting (the overlap) or being drawn from different ranges (the example described above).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

davorbonaci commented on pom.xml:
```diff
@@ -1135,6 +1136,12 @@
 
+      <dependency>
+        <groupId>io.dropwizard.metrics</groupId>
+        <artifactId>metrics-core</artifactId>
```
Generally speaking, when taking a new category of dependencies, we should do a full analysis to understand the dependency and its transitive dependencies, including license, API surface, size, bloat, etc. of all of dependencies brought in directly or transitively.

unsleepy22 commented:
*[a discussion](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1BnmeU5aYp435knSC:-Kp3e_8j-xkTAgRkRZZ5:brpy0py) (no related file):*
<details><summary><i>Previously, bjchambers (Ben Chambers) wrote…</i></summary><blockquote>

This departs from the current approach to metrics. Rather than allowing a runner to choose how to report metrics, it pushes drop wizard into the cross-runner implementation. This has other affects too, such as making it impossible to support meters across only the successfully committed work.

A lot of the complexity seems to come from the trying to use drop wizard directly while also supporting the general case of metrics with committed values if a runner supports them.

I'd like to see a discussion on the dev@ list about whether moving towards the workers reporting metrics directly to a plugin. If we like this approach, it means we can remove all the cruft around combining/storing metrics in separate cells/etc. Assuming that makes sense, we should still think about whether we want that top level connection to be made directly to drop wizard or through a layer of indirection allowing other libraries.
</blockquote></details>

Ben, Thanks for your comments. I'm ok to extend the discussion further before pulling this commit. So I'll reply on my proposal doc.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

unsleepy22 commented:
*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java, line 73 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1AXj1Ga34Inr-wHGE:-Kp3j5qqlTDPpPMWA1BI:b-3jlml2) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java#L73)):*
> support
There're two ways currently to merge metrics. The first is through Accumulators, in Flink, this works only after the job has finished (doesn't make too much sense for streaming apps IMHO), and since it's one-time-off merging, perhaps alignment doesn't matter.
The other way is forwarding beam metrics to runner metrics(current implementation) or delegating beam metrics to runner metrics(my proposal). For the former, currently we only use the count of Meter to update runner's Meter and drop m1, m5, m15 and mean rates(they're only used for Accumulators). For the latter, it's simpler because "Meter.mark" in beam calls directly to runner's "Meter.mark" and we don't even bother about merging, it's the job of external services like OpenTSDB. And it depends on metrics reporters to align the metrics.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

peihe commented:
Reviewed 3 of 4 files at r3.
Review status: 20 of 22 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pom.xml, line 1140 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp19tbbylCvL-_NJvd_:-KpDMwKL9w9yaXrwuggX:bvn6dgi) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/pom.xml#L1140)):*
<details><summary><i>Previously, bjchambers (Ben Chambers) wrote…</i></summary><blockquote>

Why are we adding dropwizard as a top-level dependency?
</blockquote></details>

I think it should be okay to add dropwizard to dependencyManagement. But, remove from the build/dependencies section below.

---

*[pom.xml, line 1141 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-KoutQhdA8m5jvY7_5JH-r4-1141:-KpDOJcMWg-iqCQpJ6CO:b-a8t796) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/pom.xml#L1141)):*
<details><summary><i>Previously, davorbonaci (Davor Bonaci) wrote…</i></summary><blockquote>

Generally speaking, when taking a new category of dependencies, we should do a full analysis to understand the dependency and its transitive dependencies, including license, API surface, size, conflicting versions with existing dependencies, bloat, etc. of all of dependencies brought in directly or transitively.

Also, new dependencies should ideally be shaded from the day-1.
</blockquote></details>

It has zero transitive dependency, so it should be okay to add (size 51KB).

https://mvnrepository.com/artifact/io.dropwizard.metrics/metrics-core/3.1.2

Leave for Cody to address shading part.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

unsleepy22 commented:
Review status: 20 of 22 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pom.xml, line 1140 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp19tbbylCvL-_NJvd_:-KpEhy8eoCG_20Ail3Z_:bftnk00) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/pom.xml#L1140)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

I think it should be okay to add dropwizard to dependencyManagement. But, remove from the build/dependencies section below.
</blockquote></details>

MeterCell depends on dropwizard Meter, that's why I added it here.

---

*[runners/core-java/pom.xml, line 99 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-KoutQhdA8m5jvY7_5JI-r4-99:-KpEiQZebVfdRWDMXFum:b-qkfosb) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/pom.xml#L99)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

I think this should live in its own module that is opt-in separate from runners-core.
</blockquote></details>

MeterCell also depends on dropwizard Meter, that's why I added it here.

---

*[runners/core-java/pom.xml, line 99 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1AA0Vvad8uhYwzMca:-KpEiUNyFCANCmSKD22r:bnw6tb8) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/pom.xml#L99)):*
<details><summary><i>Previously, bjchambers (Ben Chambers) wrote…</i></summary><blockquote>

Ditto -- dropwizard should be an implementation detail on those runners that report to dropwizard, correct?
</blockquote></details>

Yes, in the ideal way. But currently MeterCell depends on this too.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java, line 38 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1AGjycjkaup_Fr4lL:-KpEj-ceOBjtE5jtaV8Y:b5614ka) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java#L38)):*
<details><summary><i>Previously, bjchambers (Ben Chambers) wrote…</i></summary><blockquote>

Wrapping Meter like this will make it impossible for any runner to implement meters without using a drop wizard Meter. I believe that also makes it impossible to support committed values. This part of the implementation should instead capture enough information about the reported rate to export information to drop wizard within specific runners.
</blockquote></details>

It's ok for runners to implement meters without using a dropwizard Meter because my implementation only passes the Meter.count() to runners and based on the count, the runners are free to implement their own meters.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 49 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1Adm0vIK3FzCRiXe0:-KpEkDqh7BYZHM1mK8Ur:b-u1tym8) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L49)):*
<details><summary><i>Previously, bjchambers (Ben Chambers) wrote…</i></summary><blockquote>

Another similar problem -- we may get two meter data from the same thread that include overlapping rates. Again, adding the average rates together may cause numerical problems such as double counting (the overlap) or being drawn from different ranges (the example described above).
</blockquote></details>

This is a point we need to consider: In a distributed application, is the sum of `mean` rates  referable? Or can we calculate an accurate sum of mean rates? Especially for meters with different start times. I think m1,m5,m15 are more referable numbers. Even so, I still think your example is reasonable acceptable because the mean rate of each meter depicts the average rate calculated so far, i.e, an static value at 10:00, it makes sense to sum them together to get a total mean rate for all meters at that time.

For the second problem, I'm not sure about the use case, could you take an example? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

bjchambers commented:
Review status: 13 of 22 files reviewed at latest revision, 9 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1BnmeU5aYp435knSC:-KpGYG7ArVgTy9TWNjex:brshb6k) (no related file):*
<details><summary><i>Previously, unsleepy22 (Cody) wrote…</i></summary><blockquote>

Ben, Thanks for your comments. I'm ok to extend the discussion further before pulling this commit. So I'll reply on my proposal doc.
</blockquote></details>

It may also be helpful to have a quick discussion on slack to iterate on what exactly the proposal is, and then update the list. Specifically, it seems like one possible path forward is to move towards a model where all Beam metrics delegate directly to some underlying metrics implementation and then runner never gets involved.

Such a model would lose the ability to support committed metrics (we would need to make that decision consciously) but would then simplify the metrics implementations by allowing allowing the SDK to connect metrics to a metrics reporter, and then it is just a matter of implementing a Dropwizard metrics reporter. This would still mean that the Dropwizard library is only a dependency of the Dropwizard metrics reporter, not the base metrics interfaces.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java, line 38 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1AGjycjkaup_Fr4lL:-KpGXh_5IPoTAmf1e0Fn:bvshfbz) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterCell.java#L38)):*
<details><summary><i>Previously, unsleepy22 (Cody) wrote…</i></summary><blockquote>

It's ok for runners to implement meters without using a dropwizard Meter because my implementation only passes the Meter.count() to runners and based on the count, the runners are free to implement their own meters.
</blockquote></details>

The meter count isn't enough to determine the rate without knowing at what points in time those events occurred. This makes it impossible for runners to implement their own meters.

---

*[runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java, line 49 at r4](https://reviewable.io:443/reviews/apache/beam/3555#-Kp1Adm0vIK3FzCRiXe0:-KpGY7H9gXTDr74KvTmc:b5tvscc) ([raw file](https://github.com/apache/beam/blob/aa9bff62c8b88cc8c6a36ec224b799d7217dbb4d/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MeterData.java#L49)):*
<details><summary><i>Previously, unsleepy22 (Cody) wrote…</i></summary><blockquote>

This is a point we need to consider: In a distributed application, is the sum of `mean` rates  referable? Or can we calculate an accurate sum of mean rates? Especially for meters with different start times. I think m1,m5,m15 are more referable numbers. Even so, I still think your example is reasonable acceptable because the mean rate of each meter depicts the average rate calculated so far, i.e, an static value at 10:00, it makes sense to sum them together to get a total mean rate for all meters at that time.

For the second problem, I'm not sure about the use case, could you take an example? 
</blockquote></details>

I don't the sum of mean rates is meaningful, but the "total rate" does make sense. This is why just adding the means isn't enough to compute the total rate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3555)*
<!-- Sent from Reviewable.io -->

jkff commented:
Is this PR still active or should it be closed?

aaltay commented:
I will close this PR based on the [stale](https://beam.apache.org/contribute/contribution-guide/#stale-pull-requests) PR policy, please reopen if you would like to continue working on it.
